File: tier2_intermediate/dfs/solutions/p388_longest_absolute_file_path.mjs
/**
 * P388: Longest Absolute File Path [PREMIUM] (Medium)
 * https://leetcode.com/problems/longest-absolute-file-path/
 * Topics: String, Stack, Depth-First Search
 */

import { readLine, writeInt } from '../../../wasm_libs/js/io.mjs';

function solve(input) {
  const stack = [0];
  let maxLen = 0;
  const parts = input.split("\n");
  for (const part of parts) {
    let depth = 0;
    let j = 0;
    while (j < part.length && part[j] === "\t") {
      depth++;
      j++;
    }
    const name = part.slice(j);
    while (stack.length > depth + 1) {
      stack.pop();
    }
    if (name.includes(".")) {
      maxLen = Math.max(maxLen, stack[stack.length - 1] + name.length);
    } else {
      stack.push(stack[stack.length - 1] + name.length + 1);
    }
  }
  return maxLen;
}

const lines = [];
while (true) {
  const line = readLine();
  if (!line) break;
  lines.push(line);
}
const input = lines.join("\n");
writeInt(solve(input));

File: tier2_intermediate/dfs/solutions/p388_longest_absolute_file_path.py
from src.wasm_libs.py.io import read_line, write_int


def main():
    lines = []
    while True:
        line = read_line()
        if not line:
            break
        lines.append(line)
    input_str = "\n".join(lines)
    stack: list[int] = [0]
    max_len = 0
    for part in input_str.split("\n"):
        depth = part.count("\t")
        name = part.lstrip("\t")
        while len(stack) > depth + 1:
            stack.pop()
        if "." in name:
            max_len = max(max_len, stack[-1] + len(name))
        else:
            stack.append(stack[-1] + len(name) + 1)
    write_int(max_len)


if __name__ == "__main__":
    main()
